aspnetcore/grpc/versioning.md

```diff
@@ -0,0 +1,93 @@
+- **Renaming a field** - Field names are only used in generated code. The field number is used to identify fields on the network. The client will need to be updated if it upgrades to the latest contract.
+- **Renaming a message** - Message names are not sent on the network so this isn't a gRPC protocol breaking change, but the client will need to be updated if it upgrades to the latest contract.
+
+### Breaking changes
+
+These are protocol and binary breaking changes.
+
+- **Changing a field data type** - Changing a field's data type to an [incompatible type](https://developers.google.com/protocol-buffers/docs/proto3#updating) will cause errors when deserializing the message. Even if the new data type is compatible, it is likely the client will need to be updated to support the new type if it upgrades to the latest contract.
+- **Changing a field number** - The field number is used to identify fields on the network.
+- **Renaming a package, service or method** - gRPC uses the package name, service name and method name to build the URL. The client gets an *UNIMPLEMENTED* status from the server.
+- **Removing a service or method** - The client gets an *UNIMPLEMENTED* status from the server when calling the removed method.
+
+## Version number services
+
+Services should strive to remain backwards compatible with old clients. Eventually changes to your app may require breaking changes. Breaking old clients and forcing them to be updated along with your service isn't a good user experience. A way to maintain backwards compatibility while making breaking changes is to publish multiple versions of a service.
+
+gRPC supports an optional [`package`](https://developers.google.com/protocol-buffers/docs/proto3#packages) specifier, which functions much like a .NET namespace. In fact the `package` will be used as the .NET namespace for generated .NET types if `option csharp_namespace` is not set in the *.proto* file. The package can be used to specify a version number for your service and its messages:
+
+[!code-protobuf[](versioning/sample/greet.v1.proto?highlight=3)]
+
+The package name is combined with the service name to identify a service address. A service address allows multiple versions of a service to be hosted side-by-side:
+
+* `greet.v1.Greeter`
+* `greet.v2.Greeter`
+
+Implementations of the versioned service are registered in *Startup.cs*:
+
+```csharp
+app.UseEndpoints(endpoints =>
+{
+    // Implements greet.v1.Greeter
+    endpoints.MapGrpcService<GreeterServiceV1>();
+
+    // Implements greet.v2.Greeter
+    endpoints.MapGrpcService<GreeterServiceV2>();
+});
+```
+
+Including a [SemVer major version number](https://semver.org/) in the package name gives you the opportunity to publish a *v2* version of your service with breaking changes, while continuing to support older clients who call the *v1* version. Once clients have updated to use the *v2* service you can choose to remove the old version.
```

Comment:
Rather than just saying *do semver* - I'd suggest explicitly including the guidance to:
- Avoid breaking changes if reasonable
- Do not update the version unless it contains breaking changes
- Do update the version when you make breaking changes